### ISSUE TITLE
Drafts: Creating a new draft from a previous one may rearrange scenes

### ISSUE THREAD

**Issue by kevboh** [kevboh/longform#7]
See report in [discord](https://discord.com/channels/686053708261228577/805952223124520961/872467382088646767). It seems that for drafts with large files (possibly), creating via cloning rearranged scenes in both the source and destination draft.

**Comment by Moonbase59** [comment]
Example of the "reverse" modification of draft 2 of 3:

![Auswahl_181](https://user-images.githubusercontent.com/3706922/128633225-3bc6d98e-220d-471c-a13d-f5992da1f419.png)

**Comment by kevboh** [comment]
I'm starting to suspect this issue lies in the OS. I'm on macos and can't reproduce this, even with a massive draft—Longform will properly copy over the previous draft, then rearrange the new draft's scenes to match the source. The order of the originating draft is never changed; I'm not sure how it could be, looking at the code, as it's never touched.

I did notice that the copy process is much faster with Templater's new file trigger off, but it still works properly with it on.

@Moonbase59 If you can reliably reproduce this, can you do me a favor and:

1. Copy the project's `Index.md` file _before_ making a new draft from a previous one and attach it here, then
2. Make a new draft from a previous one, and
3. Copy that same file here again.

The index is the source of truth, and I'm interested in seeing how it's being mutated. Trying this on the latest version would be very helpful.

Also worth noting that I've seen weird behavior if you create new drafts from old ones while syncing, for what it's worth.

**Comment by Moonbase59** [comment]
Happy to help. Yes, I always make sure that no syncing is going on while testing. (I use Syncthing and either have all other machines off, or "pause" the synced folder that holds my Obsidian vaults. This ensures Syncthing won’t meddle.) This might be a helpful tip for other testers, though, thanks!

Testing on Linux, Obsidian 0.12.12 (AppImage), installer 0.12.10, Longform 1.0.2.
_Templater_’s "on new file creation" is _off_.

## Machine \#1: Desktop

### 1\. Old Index

I use our 200k words test again. You can see in the Index that I left _Draft 2_ mixed up (wrong order). _Draft 3_ is put into correct order by drag-n-drop in Longform.

```yaml
---
version: 1
drafts:
  - name: Draft 1
    folder: Draft 1
    scenes:
      - YAML
      - My first scene
      - My second scene
  - name: Draft 2
    folder: Draft 2
    scenes:
      - YAML
      - My first scene
      - Lorem 10k - 9
      - Lorem 10k - 8
      - Lorem 10k - 7
      - Lorem 10k - 6
      - Lorem 10k - 5
      - Lorem 10k - 4
      - Lorem 10k - 3
      - Lorem 10k - 2
      - Lorem 10k - 10
      - Lorem 10k - 1
      - Blindtext
  - name: Draft 3 200k
    folder: Draft 3 200k
    scenes:
      - YAML
      - My first scene
      - Blindtext
      - Lorem 10k - 1
      - Lorem 10k - 2
      - Lorem 10k - 3
      - Lorem 10k - 4
      - Lorem 10k - 5
      - Lorem 10k - 6
      - Lorem 10k - 7
      - Lorem 10k - 8
      - Lorem 10k - 9
      - Lorem 10k - 10
      - Lorem 10k - 11
      - Lorem 10k - 12
      - Lorem 10k - 13
      - Lorem 10k - 14
      - Lorem 10k - 15
      - Lorem 10k - 16
      - Lorem 10k - 17
      - Lorem 10k - 18
      - Lorem 10k - 19
      - Lorem 10k - 20
---

%% This file is managed by Longform. Please avoid editing it directly; doing so will almost certainly confuse the plugin, and may cause a loss of data. %%
```

### 2\. Creating new _Draft 4_

In Lonform pane: Enter `Draft 4` and select `Copy of Draft 3 200k`.

### 3\. Index directly after creating Draft 4

```yaml
---
version: 1
drafts:
  - name: Draft 1
    folder: Draft 1
    scenes:
      - YAML
      - My first scene
      - My second scene
  - name: Draft 2
    folder: Draft 2
    scenes:
      - YAML
      - My first scene
      - Lorem 10k - 9
      - Lorem 10k - 8
      - Lorem 10k - 7
      - Lorem 10k - 6
      - Lorem 10k - 5
      - Lorem 10k - 4
      - Lorem 10k - 3
      - Lorem 10k - 2
      - Lorem 10k - 10
      - Lorem 10k - 1
      - Blindtext
  - name: Draft 3 200k
    folder: Draft 3 200k
    scenes:
      - YAML
      - My first scene
      - Blindtext
      - Lorem 10k - 1
      - Lorem 10k - 2
      - Lorem 10k - 3
      - Lorem 10k - 4
      - Lorem 10k - 5
      - Lorem 10k - 6
      - Lorem 10k - 7
      - Lorem 10k - 8
      - Lorem 10k - 9
      - Lorem 10k - 10
      - Lorem 10k - 11
      - Lorem 10k - 12
      - Lorem 10k - 13
      - Lorem 10k - 14
      - Lorem 10k - 15
      - Lorem 10k - 16
      - Lorem 10k - 17
      - Lorem 10k - 18
      - Lorem 10k - 19
      - Lorem 10k - 20
  - name: Draft 4
    folder: Draft 4
    scenes:
      - YAML
      - My first scene
      - Blindtext
      - Lorem 10k - 1
      - Lorem 10k - 2
      - Lorem 10k - 3
      - Lorem 10k - 4
      - Lorem 10k - 5
      - Lorem 10k - 6
      - Lorem 10k - 7
      - Lorem 10k - 8
      - Lorem 10k - 9
      - Lorem 10k - 10
      - Lorem 10k - 11
      - Lorem 10k - 12
      - Lorem 10k - 13
      - Lorem 10k - 14
      - Lorem 10k - 15
      - Lorem 10k - 16
      - Lorem 10k - 17
      - Lorem 10k - 18
      - Lorem 10k - 19
      - Lorem 10k - 20
---


This file is managed by Longform. Please avoid editing it directly; doing so will almost certainly confuse the plugin, and may cause a loss of data.

Longform uses this file to organize your folders and notes into a project. For more details, please see [The Index File](https://github.com/kevboh/longform#the-index-file) section of the plugin’s README.
```
Looks okay.

### 4\. Restarting Obsidian

Longform pane has "jumped back" to _Draft 2_.
_Index.md_ stays the same (Draft 3 and Draft 4 not changed).

### 5\. Rearringing scenes in _Draft 4_

Moving around the scenes _Blindtext_ and _My first scene_. _Index_ follows. No mess-up.

### 6\. Compiling _Draft 4_

Doesn’t even take a second. Reult file ok, 200,626 words, 1,392,196 characters.
_Index_ still good.

### 7\. Adding a scene _Lorem 10k - 21_

For this, I change to Obsidian’s file manager, into the `Draft 4` folder, make a copy of scene `Lorem 10k - 20` and rename it to `Lorem 10k -21`.

_Index_ adds that, Longform pane shows it.

### 8\. Compile _Draft 4_again

Fast as above, result ok, _Index_ not corrupted.

### Overall result

Works as advertised. **Wonder what has happend. Maybe fixed in 1.0.2?**

**To be sure, let’s make a _Draft 5_ on another machine.

(continued)

**Comment by Moonbase59** [comment]
## Machine number 2: Laptop

Testing on Linux Mint 20.2, Obsidian 0.12.12 (AppImage), installer 0.12.10, Longform 1.0.2.
_Templater_ 1.8.1. _Templater_’s "on new file creation" is _off_.

### 0\. Syncing files over

1. Syncing switched on (Syncthing).
2. Syncs, wait until finished.
3. Syncing switched off.
4. Start Obsidian. Re-indexes because of new files added (Draft 4).
5. Look at Longform’s `Index.md` — **Already corrupt:**
   1. _Draft 3_ unchanged, _Draft 4_ rearranged.
6. In Longform’s pane, the Draft has reverted to _Draft 2_ (see #8).

### 1\. Old Index

This looks like it has to get the "new files" in, reading it in some OS sequence (inodes?), and _replacing what’s already there in `Index.md`_, thus rendering the original sequence unusable.

I wonder if Longform tries to do its thing here _before Obsidian’s indexing is finished_?

Maybe on startup it should await Obsidian’s indexer to finish (if running), _then_ check if all files in _Index_ are present, and _leave the original sequence if so_?

```yaml
---
version: 1
drafts:
  - name: Draft 1
    folder: Draft 1
    scenes:
      - YAML
      - My first scene
      - My second scene
  - name: Draft 2
    folder: Draft 2
    scenes:
      - YAML
      - My first scene
      - Lorem 10k - 9
      - Lorem 10k - 8
      - Lorem 10k - 7
      - Lorem 10k - 6
      - Lorem 10k - 5
      - Lorem 10k - 4
      - Lorem 10k - 3
      - Lorem 10k - 2
      - Lorem 10k - 10
      - Lorem 10k - 1
      - Blindtext
  - name: Draft 3 200k
    folder: Draft 3 200k
    scenes:
      - YAML
      - My first scene
      - Blindtext
      - Lorem 10k - 1
      - Lorem 10k - 2
      - Lorem 10k - 3
      - Lorem 10k - 4
      - Lorem 10k - 5
      - Lorem 10k - 6
      - Lorem 10k - 7
      - Lorem 10k - 8
      - Lorem 10k - 9
      - Lorem 10k - 10
      - Lorem 10k - 11
      - Lorem 10k - 12
      - Lorem 10k - 13
      - Lorem 10k - 14
      - Lorem 10k - 15
      - Lorem 10k - 16
      - Lorem 10k - 17
      - Lorem 10k - 18
      - Lorem 10k - 19
      - Lorem 10k - 20
  - name: Draft 4
    folder: Draft 4
    scenes:
      - YAML
      - My first scene
      - Lorem 10k - 9
      - Lorem 10k - 8
      - Lorem 10k - 7
      - Lorem 10k - 6
      - Lorem 10k - 5
      - Lorem 10k - 4
      - Lorem 10k - 3
      - Lorem 10k - 22
      - Lorem 10k - 21
      - Lorem 10k - 20
      - Lorem 10k - 2
      - Lorem 10k - 19
      - Lorem 10k - 18
      - Lorem 10k - 17
      - Lorem 10k - 16
      - Lorem 10k - 15
      - Lorem 10k - 14
      - Lorem 10k - 13
      - Lorem 10k - 12
      - Lorem 10k - 11
      - Lorem 10k - 10
      - Lorem 10k - 1
      - Blindtext
---


This file is managed by Longform. Please avoid editing it directly; doing so will almost certainly confuse the plugin, and may cause a loss of data.

Longform uses this file to organize your folders and notes into a project. For more details, please see [The Index File](https://github.com/kevboh/longform#the-index-file) section of the plugin’s README.
```
I assume this is what’s happened with the original issue, too. I’d surely have been testing Longform on both the desktop and laptop.

### Culprit found, no other tests made

Ideas?

I reckon this should be made working somehow, since people _will_ sync.

**Comment by kevboh** [comment]
Yeah, I think this is purely a sync issue. There are no hooks that I'm aware of to know when stuff is syncing. I _think_ I may have fixed this in 1.0.3 with a change to how metadata is loaded, but I'm not sure.

The reason this happens is that Longform tries to make file system changes = user changes. So, for example, creating a note in a draft without using the actual Longform UI still counts as creating a scene, and things should update accordingly. This is on purpose. What's not on purpose is that sync events (perhaps only from 3rd party solutions?) also count as "creating" a note, so Longform duly creates a scene in the order it syncs in.

**Comment by Moonbase59** [comment]
Sound logical what you explain. And yes, you won’t know when exactly sync will happen, and when it’s finished. So the situation might arise that you get "Index" way before all drafts & scenes are synced (each generating a "new file" event of course).

And yes #2, in my case the sync solution is _Syncthing_ (between Linux desktops/laptops and my Android phone).

**Comment by Edo78** [comment]
Just thinking about this issue and a possible solution could be to add frontmatter to each scene with the position.
Eg.
~~~markdown
---
lognform:
  position: 1
---
~~~
So the order in which the notes are created during sync will be irrelevant.

Obviously it's a breaking change

**Comment by kevboh** [comment]
It's not a bad idea. To be honest, I sort of want to remove this feature, and let people create new drafts like this by using native copy/paste.

**Comment by kevboh** [comment]
This is fixed in the current 2.0 beta and will reach production in the near future; see https://github.com/kevboh/longform/issues/35 for details. Closing this issue as it’s tracked in the linked one.
